src/Security/AuthenticationRequestFilter.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+namespace SilverStripe\Security;
+
+use SilverStripe\Control\HTTPResponse_Exception;
+use SilverStripe\Control\RequestFilter;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Control\Session;
+use SilverStripe\ORM\DataModel;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\ORM\ValidationException;
+
+class AuthenticationRequestFilter implements RequestFilter, IdentityStore
+{
+
+    use Configurable;
+
+    /**
+     * @return array|IdentityStore[]
+     */
+    protected function getHandlers()
```

Comment:
See my comment on security.yml; Instead of trying to construct them here, simply have getHandlers() / setHandlers(), and DI a list of named services, configured via yaml.